### ISSUE TITLE
XLA algebraic simplifier folds inverse function pairs, ignoring IEEE 754 intermediate overflow/underflow

### ISSUE THREAD

**Issue by wuyii8941** [openxla/xla#42732]
## Summary

XLA's algebraic simplifier folds inverse function compositions (`log(exp(x)) → x`, `sqrt(x²) → |x|`) even when the intermediate value overflows or underflows in float32. This changes the observable result: eager mode produces `inf`/`-inf` (correct per IEEE 754), while XLA returns a finite value.

```python
import tensorflow as tf

# exp(89.0) overflows float32 → inf. log(inf) = inf.
x = tf.constant([89.0])

# Eager: inf (correct — intermediate overflows)
print(tf.math.log(tf.exp(x)).numpy())  # [inf]

# XLA: 89.0 (WRONG — simplifier bypasses overflow)
print(tf.function(lambda x: tf.math.log(tf.exp(x)),
                  jit_compile=True)(x).numpy())  # [89.]
```

Observed on TensorFlow 2.22.0-dev20260429.

## Scope

Three inverse-pair simplifications are affected:

### 1. `log(exp(x)) → x` when exp(x) overflows (x ≥ 89 for float32)

| x | exp(x) | Eager log(exp(x)) | XLA log(exp(x)) |
|---|---|---|---|
| 88.0 | 1.65e38 (normal) | 88.0 | 88.0 | OK |
| **89.0** | **inf** (overflow) | **inf** | **89.0** | BUG |
| 100.0 | inf | inf | 100.0 | BUG |
| 200.0 | inf | inf | 200.0 | BUG |

### 2. `sqrt(x²) → |x|` when x² overflows (|x| > ~1.84e19 for float32)

| x | x² | Eager sqrt(x²) | XLA sqrt(x²) |
|---|---|---|---|
| 1e19 | 1e38 (normal) | 1e19 | 1e19 | OK |
| **1e20** | **inf** (overflow) | **inf** | **1e20** | BUG |
| -1e20 | inf | inf | 1e20 | BUG |

### 3. `exp(log(x)) → x` when x is subnormal (GPU FTZ flushes intermediate to 0)

| x | log(x) | Eager exp(log(x)) | XLA exp(log(x)) |
|---|---|---|---|
| 1e-45 | -inf (after FTZ) | 0 | 1e-45 | BUG (GPU) |
| 1e-40 | -inf (after FTZ) | 0 | 1e-40 | BUG (GPU) |

### Gradient impact

The simplification also changes gradients:

```python
# Gradient of log(exp(x)) should be 1, but at x=89:
# - Eager: NaN (exp overflows → inf, log chain rule → 0*inf = NaN)
# - XLA: 1 (simplified to identity, gradient = 1)
```

## Practical Impact

The `log(exp(x))` pattern appears when computing `log_softmax` or `logsumexp` manually:

```python
# Manual log-sum-exp without numerical stabilization:
def unstable_logsumexp(x):
    return tf.math.log(tf.reduce_sum(tf.exp(x)))

x = tf.constant([89.0, 1.0, 2.0])
# Eager: inf (correctly signals overflow in exp)
# XLA: 89.0 (simplifier hides the overflow)
```

Code that checks for inf/NaN to detect numerical issues will miss the overflow under XLA. The overflow is real — it affects the precision of the result — but XLA's simplifier silently papers over it.

The `sqrt(x²)` pattern is common in norm computations:

```python
# Euclidean norm of a vector with large elements
x = tf.constant([1e20, 1e20, 1e20])
norm = tf.sqrt(tf.reduce_sum(x ** 2))
# Eager: inf (x² overflows)
# XLA: may simplify differently
```

## Root Cause

XLA's algebraic simplifier (in `xla/service/algebraic_simplifier.cc`) applies mathematical identities:
- `log(exp(x)) → x` (since log and exp are inverses)
- `sqrt(x²) → |x|` (since sqrt and square are inverses for non-negative values)

These identities hold over the reals, but not in IEEE 754 float32 when intermediate values overflow or underflow. The simplifier should check whether the intermediate computation would overflow before applying the transformation, or guard the transformation with a range check.

## Cross-framework Check

JAX (0.9.2) has the **same bug**: `jax.jit(lambda x: jnp.log(jnp.exp(x)))(89.0) = 89.0` while eager returns `inf`. This confirms the root cause is in XLA core's algebraic simplifier.

```python
import jax, jax.numpy as jnp
f = lambda x: jnp.log(jnp.exp(x))
print(f(jnp.float32(89.0)))           # inf (eager)
print(jax.jit(f)(jnp.float32(89.0)))  # 89.0 (jit — same bug)
```

## Environment

- TensorFlow 2.22.0-dev20260429
- JAX 0.9.2: **same bug** (confirms XLA core)
- Affects both CPU and GPU XLA
- Grappler: NOT affected (does not apply this simplification)
- Forward pass: diverges (eager=inf, XLA=finite)
- Gradient: diverges (eager=NaN, XLA=1)

## Recommendation

File in `openxla/xla`. The algebraic simplifier should not fold `log(exp(x)) → x` or `sqrt(x²) → |x|` when the intermediate value provably overflows or underflows in the target float type. The simplifier already has access to the element type and could check against the overflow threshold (`log(FLT_MAX) ≈ 88.72` for float32) before applying the transformation.

**Comment by rpathade** [comment]
Hey @penumallas1261 , would love to work on this if its up for grabs

**Comment by penumallas1261** [comment]
Hi @rpathade , 

Thanks for the interest! Yes, this is absolutely up for grabs.
